src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs

```diff
@@ -0,0 +1,313 @@
+    private static int SanitizedCategoryName(Span<byte> buffer, string categoryName)
+    {
+        // Reserve 2 bytes for storing LIMIT_MAX_STR8_LENGTH_IN_BYTES and (byte)validNameLength -
+        // these 2 bytes will be back filled after iterating through categoryName.
+        int cursor = 2;
+        int validNameLength = 0;
+
+        // Special treatment for the first character.
+        var firstChar = categoryName[0];
+        if (firstChar >= 'A' && firstChar <= 'Z')
+        {
+            buffer[cursor++] = (byte)firstChar;
+            ++validNameLength;
+        }
+        else if (firstChar >= 'a' && firstChar <= 'z')
+        {
+            // If the first character in the resulting string is a lower-case alphabet,
+            // it will be converted to the corresponding upper-case.
+            buffer[cursor++] = (byte)(firstChar - 32);
+            ++validNameLength;
+        }
+        else
+        {
+            // Not a valid name.
+            return 0;
+        }
+
+        for (int i = 1; i < categoryName.Length; ++i)
+        {
+            if (validNameLength == MaxSanitizedEventNameLength)
+            {
+                break;
+            }
+
+            var cur = categoryName[i];
+            if ((cur >= 'a' && cur <= 'z') || (cur >= 'A' && cur <= 'Z') || (cur >= '0' && cur <= '9'))
+            {
+                buffer[cursor++] = (byte)cur;
+                ++validNameLength;
+            }
+        }
+
+        // Backfilling MessagePack serialization protocol and valid category length to the startIdx of the categoryName byte array.
+        MessagePackSerializer.WriteStr8Header(buffer, 0, validNameLength);
+
+        return cursor;
+    }
+
+    private byte[] ResolveTableMappingForCategoryName(string categoryName)
+    {
+        var tableNameCache = this.m_tableNameCache;
+
+        if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
+        {
+            var tableNameCache = this.m_tableNameCache;
+
+            // Check if another thread added the mapping while we waited on the
+            // lock.
+            if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+            {
+                return tableName;
+            }
+
+            byte[] mappedTableName = null;
+
+            // If user configured table name mappings run resolution logic.
+            if (this.m_tableMappings != null
+                && !this.m_tableMappings.TryGetValue(categoryName, out mappedTableName))
+            {
+                // Find best match if an exact match was not found.
+
+                string currentKey = null;
+
+                foreach (var mapping in this.m_tableMappings)
```

Comment:
This can be optimized. We can avoid checking every key of the `TableNameMappings`, if we keep a collection of the keys sorted in descending order and use that to decide if we need to keep checking for next keys.

Let's consider the example provided in the description. Let's say the category name was `MyCompany.Product2.Component` and we have the keys sorted in descending order:
```text
"MyPartner"
"MyCompany.Product2.Security"
"MyCompany.Product2"
"MyCompany.Product1"
"MyCompany"
"*"
```
Once we find a match we can break out of the loop as that would be the longest match.

Reply:
I've been working on this but it is getting kind of big with benchmarks, etc. Thinking I'll merge this PR and then open perf improvements as its own thing.